### ISSUE TITLE
Bug: False positive in 'require-atomic-updates' when adding key/value pair to object

### ISSUE THREAD

**Issue by cakoose** [eslint/eslint#17312]
### Environment

Repro in ESLint playground:
- ESLint version: 8.43.0
- Browser: Chrome 114.0.5735.133 (arm64)
- OS: macOS 13.4

### What parser are you using?

Default (Espree)

### What did you do?

ESLint playground: [link](https://eslint.org/play/#eyJ0ZXh0IjoiYXN5bmMgZnVuY3Rpb24gZihoZWFkZXJzKSB7XG4gIGhlYWRlcnNbJ2EnXSA9ICdiJztcbn1cblxuYXN5bmMgZnVuY3Rpb24gZyhoZWFkZXJzKSB7XG4gIGF3YWl0IGYoaGVhZGVycyk7XG4gIGhlYWRlcnNbJ3gnXSA9ICd5Jztcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7InJlcXVpcmUtYXRvbWljLXVwZGF0ZXMiOlsiZXJyb3IiXX0sImVudiI6eyJlczYiOnRydWV9fX0=)

```js
async function f(headers) {
  headers['a'] = 'b';
}

async function g(headers) {
  await f(headers);
  headers['x'] = 'y'; // <-- false positive?
}
```

False positive from ESLint?
> Possible race condition: `headers['x']` might be assigned based on an outdated state of `headers`.

### What did you expect to happen?

I expected no error. What's the potential race condition this is protecting against?

### What actually happened?

See above.

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiYXN5bmMgZnVuY3Rpb24gZihoZWFkZXJzKSB7XG4gIGhlYWRlcnNbJ2EnXSA9ICdiJztcbn1cblxuYXN5bmMgZnVuY3Rpb24gZyhoZWFkZXJzKSB7XG4gIGF3YWl0IGYoaGVhZGVycyk7XG4gIGhlYWRlcnNbJ3gnXSA9ICd5Jztcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7InJlcXVpcmUtYXRvbWljLXVwZGF0ZXMiOlsiZXJyb3IiXX0sImVudiI6eyJlczYiOnRydWV9fX0=

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by Rec0iL99** [comment]
Hi @cakoose, thanks for the issue. 

I don't think this is a false positive for `require-atomic-updates`. Here, `headers` is being read in `await f(headers)` and its state could've changed when the function resumes and the assignment takes place in `headers['x'] = 'y'`.

**Comment by fasttime** [comment]
Probably the same issue as #11899. The error disappears if [option `allowProperties`](https://eslint.org/docs/latest/rules/require-atomic-updates#options) is set to `true` ([**DEMO**](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IHJlcXVpcmUtYXRvbWljLXVwZGF0ZXM6IFtcImVycm9yXCIsIHsgYWxsb3dQcm9wZXJ0aWVzOiB0cnVlIH1dICovXG5cbmFzeW5jIGZ1bmN0aW9uIGYoaGVhZGVycykge1xuICBoZWFkZXJzWydhJ10gPSAnYic7XG59XG5cbmFzeW5jIGZ1bmN0aW9uIGcoaGVhZGVycykge1xuICBhd2FpdCBmKGhlYWRlcnMpO1xuICBoZWFkZXJzWyd4J10gPSAneSc7XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6eyJlczYiOnRydWV9fX0=)).

**Comment by cakoose** [comment]
> I don't think this is a false positive for `require-atomic-updates`. Here, `headers` is being read in `await f(headers)` and its state could've changed when the function resumes and the assignment takes place in `headers['x'] = 'y'` (assignment is based on a previous state of `headers`).

But isn't that just the risk with mutable parameters? For example, what about the non-async version of the same thing:

```js
function f(headers) {
  headers['a'] = 'b';
}

function g(headers) {
  f(headers);
  headers['x'] = 'y';
}
```

Would we want to report an error? It seems like the same thing?

**Comment by nzakas** [comment]
This is the expected behavior for this rule. Sync and async versions of the code are very different. Whenever you `await`, that is an opportunity for a value to change before execution resumes in that function. I know, it's a little hairy and a bit difficult to understand, but the rule is behaving correctly. As @fasttime mentioned, you can use the `allowProperties` option if you feel like there's not a risk in your specific code.

**Comment by cakoose** [comment]
Yup, for now I've set `allowProperties: true`.

The examples in the docs for require-atomic-updates ([link](https://eslint.org/docs/latest/rules/require-atomic-updates)) make sense to me, but the error in this bug report do not.

The rule's docs for properties says:

> This rule reports an assignment to a property through a variable when it detects the following execution flow in a generator or async function:
>
> 1. The variable or object property is read.
> 2. A `yield` or `await` pauses the function.
> 3. After the function is resumed, a value is assigned to a property.

Where is step 1 happening? Is it the call to `await f(headers)` that is considered a "read" of `headers`? Should that be excluded, maybe?

**Comment by Rec0iL99** [comment]
> Where is step 1 happening? Is it the call to await f(headers) that is considered a "read" of headers? Should that be excluded, maybe?

I believe yes, `headers` is being read first when passed to the function `f`.
